src/gt4py/next/iterator/transforms/cse.py

```diff
@@ -78,62 +80,109 @@ def _is_if_can_deref(node: ir.Node):
-        )
+        obj.visit(node, state=cls.State(subexprs))
         # return subexpression in pre-order of the tree, i.e. the nodes closer to the root come
         # first, and skip the root node itself
-        return {k: v for k, v in reversed(obj.subexprs.items()) if k is not node}
+        return {k: v for k, v in reversed(subexprs.items()) if k is not node}
 
     def visit(self, node, **kwargs):
-        # TODO(tehrengruber): improve this case as we might miss subexpression that could be eliminated
-        # disable collection (for all child nodes) if node matches `if_(can_deref(...), ..., ...)`
-        if _is_if_can_deref(node):
-            kwargs["allow_collection"] = False
-
         if not isinstance(node, SymbolTableTrait) and not _is_collectable_expr(node):
             return super().visit(node, **kwargs)
 
-        parents_collected_child_node_ids = kwargs.pop("collected_child_node_ids")
-        parents_used_symbol_ids = kwargs.pop("used_symbol_ids")
-        used_symbol_ids = set[int]()
+        parent_state = kwargs.pop("state")
         collected_child_node_ids = set[int]()
-        super().visit(
-            node,
-            used_symbol_ids=used_symbol_ids,
-            collected_child_node_ids=collected_child_node_ids,
-            **kwargs,
-        )
+        used_symbol_ids = set[int]()
+
+        # Special handling of `if_(can_deref(...), ..., ...)` like expressions that avoids
+        # extracting subexpressions from `if_` calls unless they are used in at least two
+        # of the three arguments.
+        if _is_if_can_deref(node):
+            # collect subexpressions for all arguments to the `if_`
+            arg_states = [self.State() for _ in range(3)]
+            for i, arg in enumerate(node.args):
+                self.visit(arg, state=arg_states[i], **kwargs)
+
+            # for each subexpression find in how many of the three arguments they occur
+            subexpr_count = {}
+            for arg_state in arg_states:
+                for subexpr in arg_state.subexprs.keys():
+                    subexpr_count.setdefault(subexpr, 0)
+                    subexpr_count[subexpr] += 1
+
+            # remove all subexpressions that are not eligible for collection
+            eligible_subexprs = {node.args[0].args[0]} | {
+                subexpr for subexpr, count in subexpr_count.items() if count >= 2
+            }
+            for arg_state in arg_states:
+                arg_state.remove_subexprs(arg_state.subexprs.keys() - eligible_subexprs)
+
+            # merge the states of the three arguments and update the state of the parent
+            new_state = self.merge_states(*arg_states)
+            for subexpr, data in new_state.subexprs.items():
+                parent_state.subexprs.setdefault(subexpr, []).extend(data)
+            collected_child_node_ids = new_state.collected_child_node_ids
+            used_symbol_ids = new_state.used_symbol_ids
```

Comment:
I think inlining the content of `merge_states` here would result in more or less the same number of lines of code and it would help to understand the algorithm (additionally, I'm still not 100% convinced that the `state.collected_child_node_ids` and `state.used_symbol_ids` fields aren't premature optimization, and seeing all the code together may help...)

Reply:
I have added todo to revisit the approach for the `collected_child_node_ids`. The `used_symbol_ids` is required and not an optimization.